src/main/java/com/example/demo/services/ItemService.java

```diff
@@ -45,9 +64,64 @@ public ItemModel getItem(int id) throws NotFoundException{
 	}
 
 	@Override
-	public ItemModel addItem(ItemModel item) {
-		// TODO Auto-generated method stub
-		return null;
+	public ItemModel addItem(ItemModel itemModel,User user) throws InvalidDataException, InsertFailedException {
+		Map<String,String> problems=(new ItemRequest(itemModel)).validate();
+		if(!problems.isEmpty()) {
+			throw new InvalidDataException(problems);
+		}
+		itemModel.setSold(false);
+		itemModel.getSeller().setId(user.getId());
+		Optional<Subcategory> subcatOpt=subcateogriesRepo.findById(itemModel.getSubcategory().getId());
+		if(subcatOpt.isEmpty()){
+			problems.clear();
+			problems.put("subcategory", "Nonexistent subcategory");
+			throw new InvalidDataException(problems);
+		}
+		
+		Item item=new Item();
+		item.populate(itemModel);
+		
+		if(item.getAddress()==null) {
+			if(user.getAddress()!=null) {
+				item.setAddress(user.getAddress());
+			}else {
+				problems.clear();
+				problems.put("address", "Address mus be defined for item or seller");
+				throw new InvalidDataException(problems);
+			}
+		}else if(item.getAddress().getId()!=0) {
+			Optional<Address> addressOpt=addressesRepo.findById(item.getAddress().getId());
+			if(addressOpt.isEmpty()) {
+				problems.clear();
+				problems.put("address", "Nonexistent address");
+				throw new InvalidDataException(problems);
+			}
+		}else {
+			item.setAddress(addressesRepo.save(item.getAddress()));
+		}
+		item=itemsRepo.save(item);
```

Comment:
since we use 2 queries to save address and to save item

what will happen if save address is successfull and save item fails

we will have bad data in our database

either move to saving only item and let address be saved in cascade

or manually add these 2 saves in SQL Transaction